compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -1344,7 +1344,31 @@ trait Applications extends Compatibility {
             unapplyArgType
           }
         val dummyArg = dummyTreeOfType(ownType)
-        val unapplyApp = typedExpr(untpd.TypedSplice(Apply(unapplyFn, dummyArg :: Nil)))
+        var unapplyApp: tpd.Tree = null
+        val unapplyAppTpe = {
+          val explicitCall = Apply(unapplyFn, dummyArg :: Nil)
+          tryEither {
+            unapplyApp = typedExpr(untpd.TypedSplice(explicitCall))
+            unapplyApp.tpe
+          } { (_, _) =>
+            // Typing explicitCall may fail if there are implicits that can only
+            // be resolved with tighter bounds on the output types
+            // (e.g. def unapply[T](x: String)(using Foo[T]): Option[T])
+            // If the call above fails to type, we try again here but with wildcards
```

Comment:
Having read through https://contributors.scala-lang.org/t/better-implicit-resolution-in-pattern-matches/5308 I sort of understand the usecase, but I'm not sure it's worth it given this kind of weird business with wildcards needed in the implementation. It seems that pattern typing is fundamentally top-down and this code attempts to work around that in some specific cases, but I can't really tell what the implications of that change is.

Reply:
Thanks for follow up. I agree that this implementation is not ideal, but I think I would disagree that pattern typing ought to be fundamentally top-down. I think it is right now, but I don't see any good reason for it. I almost filed the feature request as a bug report because it's not really clear to me whether the current behavior is the spec. 

If pattern matching were specified as syntactic sugar over a series of called to unapply/flatMap, etc., I think we might find that the behavior specified in the "feature request" would fall out naturally, and this PR would be more of a (possibly hacky) bug fix. I would like to fix the possibly hacky part, but I'll need a little consultation on what would seem less hacky and I suppose agreement on whether the syntactic sugar view of pattern matching is correct. Maybe it's worth starting a separate thread on what that view might be? Or is there already a spec for the bespoke type inference currently implemented beyond the implementation?

Reply:
> Or is there already a spec for the bespoke type inference currently implemented beyond the implementation?

The Scala 2 spec contains this section: https://www.scala-lang.org/files/archive/spec/2.13/08-pattern-matching.html#pattern-matching-expressions but I honestly have no idea how closely it matches either the Scala 2 or 3 implementation.

Reply:
That spec doesn't say anything about how to type a pattern, just how to type a pattern in the context of type variables (IIUC). And if anything, it sounds a lot like the implementation in this PR, except with undefined for type parameters instead of wildcards for implicits?
> First, it is attempted to type  [pattern] `p` with `T` as its expected type. If this fails, `p` is instead typed with a modified expected type `T'` which results from `T` by replacing every occurrence of a type parameter `a_i` by undefined.

That said, any view of pattern matching as syntactic sugar for a series of calls to `flatMap`/`orElse`/`isInstanceOf` that I could think of would not lead to the behavior that I implement in this PR, so I can't really argue this is a bug. I will close the PR for lack of interest in this new feature. Thanks for your time!